docker: add "Security" tab to image listing

larskarlitski opened the pull request:
Add a "Security" tab to the image listing, if the "org.atomic" D-Bus service exists and that particular image has been scanned already.

You cannot initiate a scan from cockpit, yet (the "Scan again" button that doesn't work).

To try it out, install atomic and a scanner, for example `atomic install fedora/atomic_scan_openscap` and run `atomic scan [image]`. The security tab should then show up for that image.

petems commented:
Awesome! Looking forward to container scanning showing up in Cockpit! 💯

larskarlitski commented:
Pushed another update. The module now polls the `org.atomic` service every ten seconds for updated vulnerability information. We can change the timeout to something higher if that makes sense and remove it completely once [atomic issue #740](https://github.com/projectatomic/atomic/issues/740) is fixed.

I've removed the "Scan again" functionality until atomic supports it (filed under the same issue).

Also, the current `atomic_dbus.py` immediately drops the `org.atomic` name, but stays running. This means that cockpit is indirectly spawning lots of those processes while polling for updates. I reported it as [atomic issue #763](https://github.com/projectatomic/atomic/issues/763) and my fix landed in master recently.

stefwalter commented:
Is this done? If so removing WIP would help get someone to review it.

larskarlitski commented:
> Is this done? If so removing WIP would help get someone to review it.

Yes, I think this is a good first step that we can already include.

There are still some test failures though, so I'm leaving needswork.

larskarlitski commented:
Rebased and fixed a whole bunch of edge cases (when atomic isn't installed or gives out inconsistent scan data). Also the `check-docker` tests that were failing should be ok again.

Also added a `warning` prop to `ListingRow`. Requested @dperpeet as a reviewer for that.

This is ready for review, so I'm removing needswork. I still need to write tests, but I'm not sure if it's ok to download the openscap container for each OS we test on.

stefwalter commented:
> This is ready for review, so I'm removing needswork. I still need to write tests, but I'm not sure if it's ok to download the openscap container for each OS we test on.

Lets test it on the atomics for now ... does that work?

dperpeet commented on pkg/docker/containers-view.jsx:
```diff
@@ -400,13 +454,29 @@ var ImageList = React.createClass({
+            var vulnerable = false;
+            var vulnerableInfo = this.state.vulnerableInfos[image.Id.replace(/^sha256:/, '')];
+            if (vulnerableInfo !== undefined) {
+                vulnerable = true; // vulnerableInfo.vulnerabilities.length;
```
I think this is a debugging change.

dperpeet commented:
Would it be a lot of work to move the changes to the listing component into a separate commit?

dperpeet commented:
Let's add an integration test also, I think testing it on the Atomics like stef suggested would be sufficient for now.

dperpeet commented:
> Would it be a lot of work to move the changes to the listing component into a separate commit?

Also, these changes should be reflected on the playground page.

larskarlitski commented:
Updated.

> I think this is a debugging change.

Thanks for the catch!

> Would it be a lot of work to move the changes to the listing component into a separate commit?

I've removed them in favor of a separate column with a badge and vulnerability count, as we discussed on Friday.

> Let's add an integration test also, I think testing it on the Atomics like stef suggested would be sufficient for now.

I started this, but atomic's D-Bus service is broken on fedora-atomic. `python-slip-dbus` is not installed.

stefwalter commented on pkg/docker/atomic.js:
```diff
@@ -0,0 +1,91 @@
+                        atomic.dispatchEvent("vulnerableInfoChanged", detailedInfos);
+                    });
+                }
+            } catch (error) {
```
This isn't okay. If there are errors to ignore ... it should be around a single specific line. For example if we want to ignore ```JSON.parse()``` errors then we shuold wrap just that line in a ```try/catch``` handler.

larskarlitski replied:
Indeed, thanks for catching this. It's a remnant of a few refactorings.

stefwalter commented on pkg/docker/containers-view.jsx:
```diff
@@ -290,7 +291,7 @@ var ImageDetails = React.createClass({
             entrypoint = image.Config.Entrypoint;
             command = image.Config.Cmd;
-            ports = image.Config.ExposedPorts || [];
+            ports = Object.keys(image.Config.ExposedPorts || {});
```
Is this change logically separate? It should be a different commit if so.

larskarlitski replied:
Yes it is. Docker gives us an Object instead of an array there. I've separated this change into its own commit.

stefwalter commented on pkg/docker/containers-view.jsx:
```diff
@@ -371,60 +417,98 @@ var ImageList = React.createClass({
+
+        var vulnerableInfo = this.state.vulnerableInfos[image.Id.replace(/^sha256:/, '')];
+        if (vulnerableInfo) {
+            var count = vulnerableInfo.vulnerabilities.length;
```
Variable hoisting is at play here. You can either use ```let``` or declare ```count``` at the function level.

larskarlitski replied:
Thanks, fixed.

stefwalter commented on pkg/docker/containers-view.jsx:
```diff
@@ -308,6 +309,50 @@ var ImageDetails = React.createClass({
+
+        } else {
+            text = cockpit.ngettext(cockpit.format('The scan from $time ($type) found one vulnerability:', args),
+                                    cockpit.format('The scan from $time ($type) found $count vulnerabilities:', args),
```
This needs to be the other way around. The arguments to ngettext should be the strings, and ```cockpit.format``` should operate on the result. At least that's how I understand it. If this does actually work, both the scanning and translating, then let me know.

stefwalter commented on pkg/docker/containers-view.jsx:
```diff
@@ -371,60 +417,98 @@ var ImageList = React.createClass({
+                    <div>
+                        <span className="pficon pficon-warning-triangle-o"></span>
+                        &nbsp;
+                        { cockpit.ngettext('1 Vulnerability', cockpit.format('$0 Vulnerabilities', count), count) }
```
This needs to be the other way around. The arguments to ngettext should be the strings, and ```cockpit.format``` should operate on the result. At least that's how I understand it. If this does actually work, both the scanning and translating, then let me know.

stefwalter commented on pkg/docker/docker.css:
```diff
@@ -1,3 +1,18 @@
+@import "/plot.css";
+@import "/term.css";
```
Out of interest, why were these two lines added here? Not necessarily a problem, but seem to be unrelated to this pull request.

larskarlitski replied:
They are indeed unnecessary.

stefwalter commented:
This is also missing documentation in:

```
doc/guide/feature-docker.xml
```

In particular the documentation for me to be able to test this should be similar to the documentation found in that file about the feature, what it calls to do the scans etc.

stefwalter commented:
There's a consistent failure like this:

```
bash: /var/lib/atomic/scan_summary.json: No such file or directory
CalledProcessError: Command 'echo '{"e02e811dd08fd49e7f6032625495118e63f597eb150403d02e3238af1df240ba": {"UUID": "e02e811dd08fd49e7f6032625495118e63f597eb150403d02e3238af1df240ba", "Scan Type": "Standards Compliance", "json_file": "/var/lib/atomic/atomic_scan_openscap/2016-12-11-12-09-57-674487/e02e811dd08fd49e7f6032625495118e63f597eb150403d02e3238af1df240ba/json", "Time": "2016-12-11T12:09:59", "Vulnerable": true, "Scanner": "openscap"}}' > /var/lib/atomic/scan_summary.json' returned non-zero exit status 1
```

stefwalter commented on pkg/docker/containers-view.jsx:
```diff
@@ -308,6 +309,50 @@ var ImageDetails = React.createClass({
+                                    'The scan from $time ($type) found $count vulnerabilities:',
+                                    info.vulnerabilities.length);
+
+            var rows = info.vulnerabilities.map(function (vulnerability) {
```
This relies on variable hoisting and is confusing. I would suggest declaring the variable outside the if block.

stefwalter replied:
Hmmm, looks like you have declared the variable above. I would just remove the 'var ' while merging ...

stefwalter commented:
I wanted to help fix this and merge it but couldn't get the tests to pass. 

 * I see this on Fedora 25: ```Error: #containers-images tr:contains("busybox:latest") td:contains("2 Vulnerabilities") not found```
 * No 'Security' tab when running Fedora 25
 * It seems like the test needs to be skipped on many other operating systems. See other tests for how to do that.

dperpeet commented on test/verify/check-docker:
```diff
@@ -498,5 +498,68 @@ CMD ["/bin/sh"]
+
+        m.execute("systemctl start docker")
+
+        # HACK put scan results for one of the images into /var/lib/atomic
```
Is there a bug/link to go with this `HACK`? So we know when we can change this / remove the hack.

larskarlitski replied:
No. It's hacky not because it's working around some issue, but because I don't yet know how we can do this in a better way. We should probably make a known-vulnerable image and scan it on every test?

larskarlitski replied:
Good point, thanks. I've changed the comment.

dperpeet commented on pkg/docker/containers-view.jsx:
```diff
@@ -319,7 +364,8 @@ var ImageList = React.createClass({
             images: [],
-            pulling: []
+            pulling: [],
+            vulnerableInfos: {}
```
Technically it's against our custom to have this info in the component state: This is information we'll probably also need to consume in other places (e.g. `X vulnerabilities found in Y images`). We can leave as is for now, but it'll probably need refactoring in the near future.

larskarlitski replied:
I don't understand. Where else would you put it?

dperpeet replied:
Usually we would separate the data store from the view, so have more data in the props. But I definitely don't think it's worth changing this now. We can have a general discussion about this sometime in the near future and consider whether your approach should become the default.

larskarlitski commented:
> I see this on Fedora 25: Error: #containers-images tr:contains("busybox:latest") td:contains("2 Vulnerabilities") not found

I didn't see this because of a race: `atomic_dbus.py` gets activated on the `org.atomic` name, but only very briefly does it actually hold on to the name. Cockpit didn't catch the `NameOwnerChanged` signal in time in all cases. If it did, it closed the channel immediately.

I worked around this by making cockpit not track the name. Thanks to @stefwalter for the idea.

> No 'Security' tab when running Fedora 25

I guess this is because of that same race. Can you please try again with the latest patch?

> It seems like the test needs to be skipped on many other operating systems. See other tests for how to do that.

Added skips for all systems in `test/README` that don't have atomic.

larskarlitski commented:
The test failure on the atomics was due to `/var/lib/atomic` missing. Just pushed a fix for that.

Note that they will still fail, because the D-Bus service doesn't work at all on atomic right now.

stefwalter commented:
> Note that they will still fail, because the D-Bus service doesn't work at all on atomic right now.

Could you include a HACK that fixes it. I think you said it's a one line python change. Otherwise we have no concrete idea this actually works once the atomic guys do a release in the next couple days.

larskarlitski commented:
> Could you include a HACK that fixes it. I think you said it's a one line python change. Otherwise we have no concrete idea this actually works once the atomic guys do a release in the next couple days.

No, that's a separate issue that I worked around with not tracking the name from cockpit.

The problem here is that the atomic is not installed correctly. `python-slip-dbus` is missing.

dperpeet commented:
fedora-atomic and rhel-atomic seem to fail the new test with the same error:
```
# testBasic (check_docker.TestAtomicScan)
#
Resource Error: Connection refused http://10.111.118.102:9090/docker
Restarting browser...
Wrote TestAtomicScan-testBasic-FAIL.png
Journal extracted to TestAtomicScan-testBasic-10.111.118.102-FAIL.log
not ok 81 testBasic (check_docker.TestAtomicScan) duration: 143s
Traceback (most recent call last):
  File "./verify/check-docker", line 567, in testBasic
    b.wait_present(row_selector + ' td span.pficon-warning-triangle-o')
  File "/build/cockpit/test/common/testlib.py", line 220, in wait_present
    return self.wait_js_func('ph_is_present', selector)
  File "/build/cockpit/test/common/testlib.py", line 214, in wait_js_func
    return self.phantom.wait("%s(%s)" % (func, ','.join(map(jsquote, args))))
  File "/build/cockpit/test/common/testlib.py", line 721, in <lambda>
    return lambda *args: self._invoke(name, *args)
  File "/build/cockpit/test/common/testlib.py", line 744, in _invoke
    raise Error(res['error'])
Error: timeout
```

larskarlitski commented:
> fedora-atomic and rhel-atomic seem to fail the new test with the same error:

No, this is the error I mentioned above. The D-Bus interface doesn't work at all on the atomics, because there's a package missing. They are aware that this is an issue. Should we blacklist the test until then?

dperpeet commented:
I can't see the security results as a regular user, I need to check the box on the login page. Can this be escalated automatically?
If it's not easily solvable, I think this is ok to do as a follow-up and shouldn't block the main part going in.

dperpeet commented:
for reference: to test this on fedora-25, I added an entry to the scans array in `/etc/atomic.d/atomic_scan_openscap`:
```
      { name: standards_compliance,
        args: ['oscapd-evaluate', 'scan', '--targets', 'chroots-in-dir:///scanin',  '--output', '/scanout', '--no-cve-scan', '-j1'],
        description: "Performs a standard scan"
      }
```
then I ran a scan with `atomic scan --scanner atomic_scan_openscap --scan_type standards_compliance docker.io/fedora:24`

dperpeet commented:
> I can't see the security results as a regular user, I need to check the box on the login page. Can this be escalated automatically?
> If it's not easily solvable, I think this is ok to do as a follow-up and shouldn't block the main part going in.

This pull request looks good, I filed two follow-up issues. If this privilege issue is easily solved then let's get that in, otherwise I'm ok with a follow-up on that also.

dperpeet commented:
Discussion on irc led to the decision to look at the privilege question as a follow-up.
I posted a fix-up pull request in #5593 with the skipped tests on rhel-atomic and fedora-atomic
